src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -374,5 +374,20 @@ public virtual async Task<VersionStamp> GetTopLevelChangeTextVersionAsync(Cancel
             var textAndVersion = await this.TextAndVersionSource.GetValueAsync(cancellationToken).ConfigureAwait(false);
             return textAndVersion.Version;
         }
+
+        public bool HasTextChanged(TextDocumentState oldState, bool ignoreUnchangeableDocument)
+        {
+            if (ignoreUnchangeableDocument && !oldState.CanApplyChange())
+            {
+                return false;
+            }
+
+            return oldState.TextAndVersionSource != TextAndVersionSource;
+        }
+
+        public bool HasInfoChanged(TextDocumentState oldState)
```

Comment:
Why adding the method? Just keeping the TextDocument vs. TextDocumentState stuff consistent? (Perfectly good reason!)

Reply:
I moved a few things around because it seems appropriate to have `TextDocument` and `TextDocumentState` answer questions about things they own instead of letting derived types do that